### ISSUE TITLE
interface-ipfs-core: timeout tests pass without implementation

### ISSUE THREAD

**Issue by koivunej** [ipfs/js-ipfs#3161]
- **Version**:

From package-lock.json:

<details><summary>interface-ipfs-core</summary>

```
"interface-ipfs-core": {
      "version": "0.137.0",
      "resolved": "https://registry.npmjs.org/interface-ipfs-core/-/interface-ipfs-core-0.137.0.tgz",
      "integrity": "sha512-//opRMeMUVnkxgkUa+ssZQh2XjhOw2stGxJsGI+ZN0d0TPsWRL3uk069eHWsQGODWo95B4663z4Y0dzxgTQQfQ==",
```
</details>

And:

<details><summary>ipfs-http-client</summary>

```
"ipfs-http-client": {
          "version": "44.0.3",
          "resolved": "https://registry.npmjs.org/ipfs-http-client/-/ipfs-http-client-44.0.3.tgz",
          "integrity": "sha512-uO/UkMryuKMqIzNkyWv056RpGZZX7UPGwIE49Y0ZAAqAmm08pxIiC1B93NP4jBefW/aNFjazcyhAewMqg1YeNw==",
          "dev": true,
          "requires": {
            "abort-controller": "^3.0.0",
            "bignumber.js": "^9.0.0",
            "bs58": "^4.0.1",
            "buffer": "^5.4.2",
            "cids": "^0.8.0",
            "debug": "^4.1.0",
            "form-data": "^3.0.0",
            "ipfs-block": "^0.8.1",
            "ipfs-core-utils": "^0.2.2",
            "ipfs-utils": "^2.2.2",
            "ipld-dag-cbor": "^0.15.1",
            "ipld-dag-pb": "^0.18.3",
            "ipld-raw": "^4.0.1",
            "iso-url": "^0.4.7",
            "it-tar": "^1.2.1",
            "it-to-buffer": "^1.0.0",
            "it-to-stream": "^0.1.1",
            "merge-options": "^2.0.0",
            "multiaddr": "^7.2.1",
            "multiaddr-to-uri": "^5.1.0",
            "multibase": "^0.7.0",
            "multicodec": "^1.0.0",
            "multihashes": "^0.4.14",
            "nanoid": "^3.0.2",
            "node-fetch": "^2.6.0",
            "parse-duration": "^0.1.2",
            "stream-to-it": "^0.2.0"
          }
        }
```
</details>

- **Platform**: Linux joonas-t495 5.3.0-62-generic #56-Ubuntu SMP Tue Jun 23 11:20:52 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

```
$ node --version
v13.14.0
```

- **Subsystem**: interface-ipfs-core

#### Description:

Run the tests before implementing the any `timeout=1ms` query parameter handling on rust-ipfs.

#### Steps to reproduce the error:

1. install rustup
2. update stable to 1.44.1
3. clone rust-ipfs, https://github.com/rs-ipfs/rust-ipfs/commit/8437373df1178dbcf55ed96745bfdb002273899a is good commit for reproducing
4. `cargo build -p http`
5. clone ipfs-rust-conformance, https://github.com/rs-ipfs/ipfs-rust-conformance/commit/1c20b41199ad594e4e8c797b70053c64121881c7 is good commit for reproducing
6. symlink the built ipfs-http binary next to `rust.sh` under ipfs-rust-conformance as `http`
7. run tests: `IPFS_RUST_EXEC="$(pwd)/rust.sh" npm test -- --grep 'should respect timeout option when'`

All tests should pass even if the functionality has not been implemented. There are probably issues with following the above instructions, so no harm in not trying. Good to print them out for once at least :)

I suspect that ipfs-http-client library internally handles the timeout while also rendering it to the query parameter. Perhaps the http client level timeout should be disabled when doing these timeout tests? If there is such way in the ipfs-http-client, I couldn't find it in https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs-http-client/README.md#global-timeouts.

**Comment by achingbrain** [comment]
I think this is a hangover from before timeouts were implemented in js-ipfs, so the client also has a timeout.

> Perhaps the http client level timeout should be disabled when doing these timeout tests?

Sounds good, we can add an option to the http client.

**Comment by achingbrain** [comment]
This might be better done if we just remove the [default timeout](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs-http-client/src/lib/core.js#L113).

@koivunej could you try removing the ` || 60000 * 20` from your local copy?

I'm trying to follow your instructions but I'm getting an error `error: package ID specification `http` matched no packages` during the `cargo build -p http` step.  I notice the `ipfs-rust-conformance` module has been archived so I guess the instructions won't repro any more.

**Comment by koivunej** [comment]
@achingbrain I applied this:

```diff
@@ -110,7 +110,7 @@
     /** @type {ClientOptions} */
     const opts = normalizeInput(options)
     super({
-      timeout: parseTimeout(opts.timeout) || 60000 * 20,
+      timeout: parseTimeout(opts.timeout),
       headers: opts.headers,
       base: normalizeInput(opts.url).toString(),
       handleError: errorHandler,
```

With this patch and how we currently run the interface-http-core tests **now in [rust-ipfs/conformance](https://github.com/rs-ipfs/rust-ipfs/tree/master/conformance)** (we simplified from two repos to just one), I cannot see any difference in outcome before or after removing the timeout implementation from `dag/resolve` and `block/get` while passing the tests.

Tried also adding an "assertion" that `opts.timeout` is never truthy after `normalizeInput(options)`, and it's never truthy.

I got a single random test failure when running `npm test -- --grep timeout` with the "should respect timeout option when getting a DAG node" as follows:

```
  1) .dag.get
       should respect timeout option when getting a DAG node:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (.../rust-ipfs/conformance/test/index.js)
      at listOnTimeout (internal/timers.js:549:17)
      at processTimers (internal/timers.js:492:7)
```

When running the test in loop with:

```
$ while true; do IPFS_RUST_EXEC="$(pwd)/rust.sh" npm test -- --grep 'should respect timeout option when getting a DAG node'; done
```

It passes every time in 300-400ms:

```
> ipfs-rust-conformance@1.0.0 test /home/joonas/src/eq/public/ipfs-rust/rust-ipfs/conformance
> npx mocha "--grep" "should respect timeout option when getting a DAG node"


  .dag.get
    ✓ should respect timeout option when getting a DAG node


  1 passing (389ms)
```

If it passes normally in 300-400ms and the timeout is 2000ms, I would expect "not implemented ?timeout=" to consistently hit the 2s timeout, which it certainly would do in a manual test with curl:

```
$ curl '127.0.0.1:41233/api/v0/dag/resolve?arg=/ipfs/QmUm3BmzTKE1ncMiKJ5EKA5uVMamiJiy3mGxopGAA12hvX&timeout=1ms'
```

This hangs forever as expected with the timeout parameter processing commented out.

**Comment by koivunej** [comment]
Fixed repro instructions (I had used wrong name for ipfs-http of course :):

1. install rustup
2. update stable to 1.44.1
3. clone rust-ipfs, rs-ipfs/rust-ipfs@2576a6d77a2e73b30f4690cc5b2a340e90f5ec15 (latest master) is good commit for reproducing
4. cargo build -p ipfs-http
5. cd conformance && ./setup.sh
6. run tests: IPFS_RUST_EXEC="$(pwd)/rust.sh" npm test -- --grep 'should respect timeout option when'
